shared/subspace-kzg/src/tests.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{Kzg, Scalar};
+use rand::thread_rng;
+use rand_core::RngCore;
+use subspace_core_primitives::ScalarBytes;
+
+#[test]
+fn basic() {
+    let values = (0..8)
+        .map(|_| Scalar::from(rand::random::<[u8; ScalarBytes::SAFE_BYTES]>()))
+        .collect::<Vec<_>>();
+
+    let kzg = Kzg::new();
+    let polynomial = kzg.poly(&values).unwrap();
+    let commitment = kzg.commit(&polynomial).unwrap();
+
+    let num_values = values.len();
+
+    for (index, value) in values.iter().enumerate() {
+        let index = index.try_into().unwrap();
+
+        let witness = kzg.create_witness(&polynomial, num_values, index).unwrap();
+
+        assert!(
+            kzg.verify(&commitment, num_values, index, value, &witness),
+            "failed on index {index}"
+        );
+    }
+}
+
+#[test]
+fn bytes_scalars_conversion() {
```

Comment:
Because of the safety assumption around the layout of `Option<Scalar>`, it would be good to have a specific test for `Vec<Option<Scalar>>` conversion (and the reverse) here.

Reply:
Added constant assertions about size and alignment instead, let me know if that is what you meant

Reply:
In this specific case, I agree that’s enough to detect all potential variations, including field order (which is odd, but specifically documented as something the compiler can do if there are different optimisation settings for different crates). Because the data field has alignment requirements, putting the discriminant before the data would add internal padding, and change the size of the option.

As a counter example, it wouldn’t work for `Option<u8>`, where we’d need a runtime test that the data bytes matched.

Reply:
While what you described makes sense, it'd still be odd for compiler to put discriminant in one location for inner struct and another location for its transparent wrapper. I acknowledge it is technically allowed to do that, but still I'd expect it to be consistent. Thankfully we have a larger alignment here, so we're lucky I guess.

Interestingly, there should be two bits inside of this value that could have been used as a niche, which would allow us to have more control over `Option` behavior, but maybe at some point in the future :roll_eyes: